### PR TITLE
Fix bugs in model causing it to return partial and incorrect results.

### PR DESCRIPTION
**Pull request by djsauble**
This PR makes the cosine similarity model actually useful!

* Results were wrong because we didn't transform package identifiers back to their original values after running everything through the count vectorizer and computing the similarity matrix. Thus recommendations were mostly random and not useful.
* We were also only getting half the expected results because the similarity matrix was only non-zero for the upper half of the matrix. This PR mirrors the results and appends to the database so we get the full similarity matrix now.

For comparison, before we got ~2,000 package recommendations for express. Now we get ~8,000 recommendations. To keep performance at reasonable levels, I've limited the number of results to 1,000, but we should consider ways to raise this limit (e.g. server-side pagination).

### CODE REVIEW

File: webserver/pkgpkr/webservice/recommender_service.py [hunk 1]
```diff
@@ -47,17 +47,23 @@ def get_recommendations(self, dependencies):
...
                     INNER JOIN packages b ON s.package_b = b.id
-                    WHERE s.package_a IN (
-                        SELECT DISTINCT id FROM packages WHERE name in ({str(packages)[1:-1]})
-                    )
+                    WHERE
+                    s.package_a IN (SELECT id FROM packages WHERE name in ({str(packages)[1:-1]}))
+                    AND
+                    s.package_b NOT IN (SELECT id FROM packages WHERE name in ({str(packages)[1:-1]}))
+                    ORDER BY b.name, s.similarity DESC
+                    LIMIT 1000
```
Comment (vpopil) [inline webserver/pkgpkr/webservice/recommender_service.py:66]: maybe have 1000 as param here?

File: pipeline/scraper/npm_query.py [hunk 2]
```diff
@@ -35,25 +35,34 @@ def get_package_metadata(dependency):
...
-        downloads_last_month = get_average_monthly_donwloads(res_json.get('downloads'))
-        entry['downloads_last_month'] = downloads_last_month
+    try:
+        res = requests.get(f'{NPM_LAST_MONTH_DOWNLOADS_META_API_URL}/{dependency_name}')
+        res_json = res.json()
+
+        entry['downloads_last_month'] = 0
+        if res_json.get('downloads'):
+            downloads_last_month = get_average_monthly_donwloads(res_json.get('downloads'))
+            entry['downloads_last_month'] = downloads_last_month
```
Comment (vpopil) [inline pipeline/scraper/npm_query.py:45]: maybe no need for interim results

File: pipeline/model/generate.py [hunk 3]
```diff
@@ -52,7 +56,17 @@
...
+
+# Map the package identifiers back to their pre-vectorized values
+MAPPING = create_map([lit(x) for x in chain(*enumerate(VECTORIZER_MODEL.vocabulary))])
+SIMILARITY_DF = SIMILARITY_DF.withColumn("package_a", MAPPING.getItem(col("a")).cast("integer")) \
+                             .withColumn("package_b", MAPPING.getItem(col("b")).cast("integer"))
+SIMILARITY_DF = SIMILARITY_DF.drop(col("a")).drop(col("b"))
+
+# Mirror the columns and append to the existing dataframe so we need only query the first column
+SIMILARITY_DF = SIMILARITY_DF.select('package_a', 'package_b', 'similarity') \
+                             .union(SIMILARITY_DF.select('package_b', 'package_a', 'similarity'))
```
Comment (vpopil) [inline pipeline/model/generate.py:69]: very nice

File: pipeline/model/generate.py [hunk 4]
```diff
@@ -9,8 +9,12 @@
 from pyspark.ml.feature import CountVectorizer
 from pyspark.sql.functions import col
 from pyspark.sql.functions import collect_list
+from pyspark.sql.functions import create_map
+from pyspark.sql.functions import lit
```
[the remark this answers is not in the document]
Comment (djsauble) [reply pipeline/model/generate.py]: Done

File: pipeline/scraper/npm_query.py (hunk 2 above)
Comment (djsauble) [reply pipeline/scraper/npm_query.py:45]: Fixed

File: pipeline/scraper/npm_query.py [hunk 5]
```diff
@@ -35,25 +35,34 @@ def get_package_metadata(dependency):
...
+        res = requests.get(f'{NPM_DEPENDENCY_META_URL}/{dependency_name}')
+        res_json = res.json()
+
+        entry['categories'] = None
+        if res_json.get('keywords'):
+            entry['categories'] = res_json.get('keywords')
+
+        entry['modified'] = None
+        if res_json.get('time') and res_json.get('time')['modified']:
+            entry['modified'] = res_json.get('time')['modified']
```
[the remark this answers is not in the document]
Comment (djsauble) [reply pipeline/scraper/npm_query.py:60]: Fixed

File: webserver/pkgpkr/webservice/recommender_service.py (hunk 1 above)
Comment (djsauble) [reply webserver/pkgpkr/webservice/recommender_service.py:66]: Yeah, I'll pull it out into a global variable. Don't really want to add yet another environment variable though.

**Pull request merged**